File: apps/android/app/src/main/java/org/peerscholar/app/data/AppStore.kt
package org.peerscholar.app.data

import android.content.Context
import android.content.SharedPreferences
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.json.JSONObject

enum class SyncStatus(val label: String) {
    LOCAL("Saved on this device"),
    SYNCING("Syncing…"),
    SYNCED("Synced to your account"),
    ERROR("Saved locally — sync unavailable"),
}

data class LearnerUiState(
    val enrollments: Set<String> = emptySet(),
    val completedLessons: Map<String, Set<String>> = emptyMap(),
    val bookings: Set<String> = emptySet(),
    val syncStatus: SyncStatus = SyncStatus.LOCAL,
)

/**
 * Learner state for the Android app: enrollments, per-lesson completion, and
 * session bookings.
 *
 * Mirrors the web app's AppStore (apps/web/lib/AppStore.tsx) and the iOS
 * AppStore so all three front ends behave the same. Persists to
 * SharedPreferences so the flow works signed-out and offline; LearnerSync
 * layers Firestore on top when someone is signed in.
 */
class AppStore : ViewModel() {
    private val _state = MutableStateFlow(LearnerUiState())
    val state: StateFlow<LearnerUiState> = _state

    private var prefs: SharedPreferences? = null

    fun attach(context: Context) {
        if (prefs != null) return
        prefs = context.applicationContext.getSharedPreferences("peerscholar", Context.MODE_PRIVATE)
        load()
    }

    // ---- Enrollment ----

    fun isEnrolled(courseId: String) = _state.value.enrollments.contains(courseId)

    fun enroll(courseId: String) {
        if (isEnrolled(courseId)) return
        _state.value = _state.value.copy(enrollments = _state.value.enrollments + courseId)
        persist()
    }

    fun unenroll(courseId: String) {
        _state.value = _state.value.copy(enrollments = _state.value.enrollments - courseId)
        persist()
    }

    // ---- Lesson progress ----

    fun isLessonComplete(courseId: String, lessonId: String) =
        _state.value.completedLessons[courseId]?.contains(lessonId) == true

    fun toggleLesson(courseId: String, lessonId: String) {
        val done = _state.value.completedLessons[courseId] ?: emptySet()
        val next = if (done.contains(lessonId)) done - lessonId else done + lessonId
        _state.value = _state.value.copy(
            completedLessons = _state.value.completedLessons + (courseId to next)
        )
        persist()
    }

    fun completedCount(courseId: String) = _state.value.completedLessons[courseId]?.size ?: 0

    /** Progress as 0f..1f so it can drive LinearProgressIndicator directly. */
    fun progress(courseId: String, totalLessons: Int): Float =
        if (totalLessons <= 0) 0f else completedCount(courseId).toFloat() / totalLessons

    // ---- Bookings ----

    fun isBooked(sessionId: String) = _state.value.bookings.contains(sessionId)

    fun book(sessionId: String) {
        if (isBooked(sessionId)) return
        _state.value = _state.value.copy(bookings = _state.value.bookings + sessionId)
        persist()
    }

    fun cancelBooking(sessionId: String) {
        _state.value = _state.value.copy(bookings = _state.value.bookings - sessionId)
        persist()
    }

    // ---- Persistence ----

    private fun load() {
        val p = prefs ?: return
        val enrollments = p.getStringSet(KEY_ENROLLMENTS, emptySet())?.toSet() ?: emptySet()
        val bookings = p.getStringSet(KEY_BOOKINGS, emptySet())?.toSet() ?: emptySet()
        val lessons = decodeLessons(p.getString(KEY_LESSONS, null))
        _state.value = _state.value.copy(
            enrollments = enrollments,
            bookings = bookings,
            completedLessons = lessons,
        )
    }

    private fun persist() {
        val p = prefs ?: return
        p.edit()
            .putStringSet(KEY_ENROLLMENTS, _state.value.enrollments)
            .putStringSet(KEY_BOOKINGS, _state.value.bookings)
            .putString(KEY_LESSONS, encodeLessons(_state.value.completedLessons))
            .apply()
        pushToCloud()
    }

    // SharedPreferences has no nested-map type, so the per-course lesson sets
    // are stored as a small JSON blob.
    private fun encodeLessons(map: Map<String, Set<String>>): String {
        val root = JSONObject()
        map.forEach { (courseId, lessons) -> root.put(courseId, lessons.joinToString(",")) }
        return root.toString()
    }

    private fun decodeLessons(raw: String?): Map<String, Set<String>> {
        if (raw.isNullOrBlank()) return emptyMap()
        return runCatching {
            val root = JSONObject(raw)
            buildMap {
                root.keys().forEach { key ->
                    val joined = root.optString(key, "")
                    put(key, if (joined.isBlank()) emptySet() else joined.split(",").toSet())
                }
            }
        }.getOrDefault(emptyMap())
    }

    // ---- Firestore sync ----

    /**
     * Pulls cloud state and unions it with what's on device, then writes the
     * result back — so signing in on a second device adds to your history
     * rather than clobbering either side.
     */
    fun syncOnSignIn() {
        val uid = LearnerSync.currentUid() ?: return
        _state.value = _state.value.copy(syncStatus = SyncStatus.SYNCING)
        viewModelScope.launch {
            runCatching {
                val remote = LearnerSync.fetch(uid)
                val mergedLessons = _state.value.completedLessons.toMutableMap()
                remote.completedLessons.forEach { (courseId, lessons) ->
                    mergedLessons[courseId] = (mergedLessons[courseId] ?: emptySet()) + lessons
                }
                _state.value = _state.value.copy(
                    enrollments = _state.value.enrollments + remote.enrollments,
                    bookings = _state.value.bookings + remote.bookings,
                    completedLessons = mergedLessons,
                )
                prefs?.edit()
                    ?.putStringSet(KEY_ENROLLMENTS, _state.value.enrollments)
                    ?.putStringSet(KEY_BOOKINGS, _state.value.bookings)
                    ?.putString(KEY_LESSONS, encodeLessons(_state.value.completedLessons))
                    ?.apply()
                LearnerSync.save(uid, snapshot())
            }.onSuccess {
                _state.value = _state.value.copy(syncStatus = SyncStatus.SYNCED)
            }.onFailure {
                // Rules not deployed, offline, or Firebase absent — the local
                // layer already holds everything, so degrade rather than fail.
                _state.value = _state.value.copy(syncStatus = SyncStatus.ERROR)
            }
        }
    }

    fun signedOut() {
        _state.value = _state.value.copy(syncStatus = SyncStatus.LOCAL)
    }

    private fun pushToCloud() {
        val uid = LearnerSync.currentUid() ?: return
        if (_state.value.syncStatus == SyncStatus.SYNCING) return
        viewModelScope.launch {
            runCatching { LearnerSync.save(uid, snapshot()) }
                .onSuccess { _state.value = _state.value.copy(syncStatus = SyncStatus.SYNCED) }
                .onFailure { _state.value = _state.value.copy(syncStatus = SyncStatus.ERROR) }
        }
    }

    private fun snapshot() = LearnerSync.State(
        enrollments = _state.value.enrollments.toList(),
        completedLessons = _state.value.completedLessons.mapValues { it.value.toList() },
        bookings = _state.value.bookings.toList(),
    )

    private companion object {
        const val KEY_ENROLLMENTS = "enrollments"
        const val KEY_BOOKINGS = "bookings"
        const val KEY_LESSONS = "completedLessons"
    }
}

File: apps/ios/PeerScholar/ViewModels/AppStore.swift
import Foundation
import Combine

/// Learner state for the iOS app: enrollments, per-lesson completion, and
/// session bookings.
///
/// Mirrors the web app's AppStore (apps/web/lib/AppStore.tsx) so both front
/// ends behave the same. Persists to UserDefaults, which keeps the whole flow
/// working signed-out and offline; `LearnerSync` layers Firestore on top when
/// someone is signed in.
@MainActor
final class AppStore: ObservableObject {
    @Published private(set) var enrollments: Set<String> = []
    @Published private(set) var completedLessons: [String: Set<String>] = [:]
    @Published private(set) var bookings: Set<String> = []
    @Published private(set) var syncStatus: SyncStatus = .local

    enum SyncStatus {
        case local, syncing, synced, error

        var label: String {
            switch self {
            case .local: return "Saved on this device"
            case .syncing: return "Syncing…"
            case .synced: return "Synced to your account"
            case .error: return "Saved locally — sync unavailable"
            }
        }
    }

    private let defaults = UserDefaults.standard
    private let enrollmentsKey = "peerscholar.enrollments"
    private let lessonsKey = "peerscholar.completedLessons"
    private let bookingsKey = "peerscholar.bookings"

    init() {
        load()
    }

    // MARK: - Enrollment

    func isEnrolled(_ courseId: String) -> Bool { enrollments.contains(courseId) }

    func enroll(_ courseId: String) {
        guard !enrollments.contains(courseId) else { return }
        enrollments.insert(courseId)
        persist()
    }

    func unenroll(_ courseId: String) {
        enrollments.remove(courseId)
        persist()
    }

    // MARK: - Lesson progress

    func isLessonComplete(courseId: String, lessonId: String) -> Bool {
        completedLessons[courseId]?.contains(lessonId) ?? false
    }

    func toggleLesson(courseId: String, lessonId: String) {
        var done = completedLessons[courseId] ?? []
        if done.contains(lessonId) {
            done.remove(lessonId)
        } else {
            done.insert(lessonId)
        }
        completedLessons[courseId] = done
        persist()
    }

    func completedCount(courseId: String) -> Int {
        completedLessons[courseId]?.count ?? 0
    }

    /// Progress as 0...1 so it can drive ProgressView directly.
    func progress(courseId: String, totalLessons: Int) -> Double {
        guard totalLessons > 0 else { return 0 }
        return Double(completedCount(courseId: courseId)) / Double(totalLessons)
    }

    // MARK: - Bookings

    func isBooked(_ sessionId: String) -> Bool { bookings.contains(sessionId) }

    func book(_ sessionId: String) {
        guard !bookings.contains(sessionId) else { return }
        bookings.insert(sessionId)
        persist()
    }

    func cancelBooking(_ sessionId: String) {
        bookings.remove(sessionId)
        persist()
    }

    // MARK: - Persistence

    private func load() {
        enrollments = Set(defaults.stringArray(forKey: enrollmentsKey) ?? [])
        bookings = Set(defaults.stringArray(forKey: bookingsKey) ?? [])
        if let raw = defaults.dictionary(forKey: lessonsKey) as? [String: [String]] {
            completedLessons = raw.mapValues(Set.init)
        }
    }

    private func persist() {
        defaults.set(Array(enrollments), forKey: enrollmentsKey)
        defaults.set(Array(bookings), forKey: bookingsKey)
        defaults.set(completedLessons.mapValues(Array.init), forKey: lessonsKey)
        Task { await pushToCloud() }
    }

    // MARK: - Firestore sync

    /// Pulls cloud state and unions it with what's on device, then writes the
    /// result back. Union semantics mean signing in on a second device adds to
    /// your history rather than clobbering either side.
    func syncOnSignIn(uid: String) async {
        syncStatus = .syncing
        do {
            let remote = try await LearnerSync.fetch(uid: uid)
            enrollments.formUnion(remote.enrollments)
            bookings.formUnion(remote.bookings)
            for (courseId, lessons) in remote.completedLessons {
                completedLessons[courseId, default: []].formUnion(lessons)
            }
            defaults.set(Array(enrollments), forKey: enrollmentsKey)
            defaults.set(Array(bookings), forKey: bookingsKey)
            defaults.set(completedLessons.mapValues(Array.init), forKey: lessonsKey)

            try await LearnerSync.save(uid: uid, state: snapshot())
            syncStatus = .synced
        } catch {
            // Rules not deployed, offline, or Firebase absent — the local layer
            // already holds everything, so degrade rather than fail.
            syncStatus = .error
        }
    }

    func signedOut() {
        syncStatus = .local
    }

    private func pushToCloud() async {
        guard let uid = LearnerSync.currentUid, syncStatus != .syncing else { return }
        do {
            try await LearnerSync.save(uid: uid, state: snapshot())
            syncStatus = .synced
        } catch {
            syncStatus = .error
        }
    }

    private func snapshot() -> LearnerSync.State {
        LearnerSync.State(
            enrollments: Array(enrollments),
            completedLessons: completedLessons.mapValues(Array.init),
            bookings: Array(bookings)
        )
    }
}
